kart/spatial_filters.py

```diff
@@ -0,0 +1,117 @@
+import collections
+import functools
+import logging
+
+import click
+
+from .geometry import Geometry
+
+L = logging.getLogger("kart.spatial_filters")
+
+
+class SpatialFilter:
+    @classmethod
+    @functools.lru_cache()
+    def from_spec(cls, spec):
+        if not spec:
+            return SpatialFilter.MATCH_ALL
+
+        # Could also accept WKT here if that is useful.
+        filter_geometry = Geometry.from_hex_wkb(spec).normalise()
+        return SpatialFilter(filter_geometry)
+
+    def __init__(self, filter_geometry, match_all=False):
```

Comment:
ugh, I had assumed we could just avoid dealing with CRSes at all and just assume the filter matches the dataset. But just realised that since a repo can have multiple datasets in different CRSes, and a filter is at the _repo_ level, that might not work well.

But dealing with transforming these things means we have to deal with:
* normalisation of geographic things (potentially both features and filters) wrt the antimeridian
* filters that can't be [fully] projected into the dataset CRS (but some features might still match the filter)
* features that can't be [fully] projected into the filter CRS (but might still match the filter)
* interpolation of long unnoded lines before reprojection
all of that feels like it could easily be as much code as this entire PR so far :/

Or we could ignore it all and deal with it down the road a bit...

Reply:
Can be punted as long as we have an overall outline/approach we can keep fleshing out.

Reply:
Here is an approach - for each dataset, transform the filter to the dataset's CRS, making sure we "interpolate long unnoded lines before the reprojection". Remove those parts of the filter that don't fit in the target CRS. Add two more filters identical to the first, but one 360 degrees East of it and one 360 degrees West of it, to make sure we features that only match when everything normalised. (Remove all or part of these too if they don't fit in the target CRS). Generate some envelopes that contain the filter (3 seems like a good number, but could be more or less). Then, as now, test every feature envelope with the filter envelope(s), and if they match, test the feature geometry against the filter geometry(s).

This approach may need some changes, since I've never done something like this before, but critically it doesn't need any changes to the current code or data structures - we could do this work at any time. I think this is true in general - whatever approach we take, we should be able to swap it in later without having to restructure all of Kart.

The other thing to consider is whether we can convert any spatial filter to a set of S2 cells - if we will hit troubles there. This seems like it should be okay too though: any geometry with any CRS should be convertable to a WGS84 geometry and from there to S2 cells